### PR TITLE
adding scPRINT and its suite of packages

### PR DESCRIPTION
**Pull request by jkobject**
> Sorry to add 4 tools at once but they are quite related, and it seems simpler to review them all at once instead of making 4 distinct but related PRs 

Name of the tool:
- scPRINT
- scDataLoader
- benGRN
- GRnnData

Short description:
- A single-cell foundation model for Gene Network inference and more.
- A dataloader to work with large single-cell databases.
- A benchmarking tool for gene network inference from scRNAseq.
- Gene Network-enhanced AnnData with utils and conversions.

How does the package use scverse data structures (please describe in a few sentences): 
- scPRINT can be called directly in command line or in python on an anndata object. it also outputs updated anndatas. 
- scDataLoader can work on one to many h5ad/zarr saved anndatas and dataload them in parallel. It serves as an alternative and might be more efficient than anndata's dataloader.
- GRnnData is AnnData with some more utils and loading mechanism when interfacing with tools producing gene networks. We made it when stumbling upon the many ways people were using gene networks in single cell omics
- benGRN is the "benchmarking side" of GRnnData. It is like scIB but for gene network inference instead of embeddings.

### mandatory

-   [x] The code is publicly available under an [OSI-approved](https://opensource.org/licenses/alphabetical) license
-   [x] The package provides versioned releases
-   [x] The package can be installed from a standard registry (e.g. PyPI, conda-forge, bioconda)
-   [x] Automated tests cover essential functions of the package and a reasonable range of inputs and conditions [^1]
-   [x] Continuous integration (CI) automatically executes these tests on each push or pull request [^2]
-   [x] The package provides API documentation via a website or README[^3]
-   [x] The package uses scverse datastructures where appropriate (i.e. AnnData, MuData or SpatialData and their modality-specific extensions)
-   [x] I am an author or maintainer of the tool and agree on listing the package on the scverse website

### Recommended

-   [x] Please announce this package on scverse communication channels (zulip, discourse, twitter)
-   [x] Please tag the author(s) these announcements. Handles (e.g. `@scverse_team`) to include are:
    -   Twitter: @jkobject
    -   Zulip:
    -   Discourse:
    -   Mastodon:
-   [x] The package provides tutorials (or "vignettes") that help getting users started quickly
-   [ ] The package uses the [scverse cookiecutter template](https://github.com/scverse/cookiecutter-scverse).
   -  It is using Rochard's project template

[^1]: We recommend thtat tests cover at least all user facing (public) functions. Minimal tests ensure that the function does not fail on an example data set. Ideally, tests also ensure the correctness of the results, e.g. by comparing against a snapshot.

[^2]: Continuous integration means that software tests are automatically executed on every push to the git repository. This guarantees they are always run and that they are run in a clean environment. Scverse ecosystem packages most commonly use [GitHub Actions](https://github.com/features/actions) for CI. For an example, check out our [cookiecutter template](https://github.com/scverse/cookiecutter-scverse).

[^3]: By API documentation, we mean an overview of _all_ public functions provided a package, with documentation of their parameters. For an example, see the [Scanpy documentation](https://scanpy.readthedocs.io/en/stable/api/preprocessing.html). In simple cases, this can be done manually in a README file. For anything more complex, we recommend the [Sphinx Autodoc plugin](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html)

### CODE REVIEW

**Comment by jkobject** [comment]
Hello, do you have any feedback on this?

**Comment by grst** [comment]
Sorry, I'm a bit behind on this due to vacation time. 
I'll review it beginning of October.

**Comment by grst** [comment]
Hi @jkobject,

thanks for submitting the packages. Overall, this looks good, my main concern is test coverage. According to our guidelines we require that
> Automated tests cover essential functions of the package and a reasonable range of inputs and conditions

which is not the case for any of the packages. Really most/all functions of your public API should be covered and tests should ideally also test for correct outputs and not just that no Exception occurs. 


### Other feedback
 * The documentation links in the yaml file (e.g. https://jkobject.com/GRnnData) don't work for me
 * Consider renaming the header "Documentation" to "API documentation" or "API", it took me a while to find the API docs
image

**Comment by jkobject** [comment]
Hi @grst,

Thanks for the feedback! I do understand that coverage should be maximal and will update the codes to test more files. Many files are actually barely used anymore and might be better removed I believe. 

For scPRINT, while I test the denoising pipeline and make sure that it is performing well. I might not be able to test the quality of the results as easily for the network inference, embedding, and classification part. But I will add more tests than just exceptions. I will come back to you in a few weeks with the updates!

Best,

**Comment by jkobject** [comment]
Hello @grst,

I made a lot of updates to all packages and updated the .yml files.

1. Now benGRN reaches 53% code coverage. Now all functions are tested although not every parameters (mostly the parts that run the same thing with variations)
Also, GENIE3 is now tested, but many subcases are not tested. For using the tool this should cover most usecases. 

2. Now grnndata reaches 66% code coverage. Only 2 edgecases of loading loom scope files and plotting an HTML version of the graph is not tested yet.

3. Now scDataloader reaches 52% codecoverage. Again mostly due to the many different possible options for preprocessing datasets and generating a minibatch. But it better covers the base use-cases, too, now.

4. Now scPRINT reaches 33% code coverage. I now test all inference usecases. However scPRINT is a foundation model and can be trained. I can try to design a simple, super small model training but this requires specific kinds of GPUs with specific CUDA versions and I am not sure that most foundation models undergo automated testing this way. Moreover the rest of untested scripts are also related to training and to downloading of sequence data and running ESM2. 
Finally many options can be provided to run the model and I don't test them all, especially for gene network inference. This is also for the sake of time as otherwise the tests might take more than half an hour to run. 

Let me know if this works with you or if I need to go further for the automated testing! 

Best,

**Comment by jkobject** [comment]
You are right and it is in my to do list to test the fit function on a toy dataset. It seemed a bit more work than the other tests that is all. I would want to get the loss and check that it is going down. However it won't assess all the triton functions that I got from flashattention2 and openai. Since they need specific GPUs..

I have a bit of time today. I will try to implement and let you know if it is a quick update or more involved.

**Review by grst — APPROVED** [review]
